062-support-tiered-storage.md

```diff
@@ -0,0 +1,92 @@
+
+The RSM config only supports custom type. The config allows users to specify the RSM class information.
+
+Because the `RemoteStorageManager` dependency is provided by Kafka plugin, the implementation could vary case by case. It’s important to allow user to pass in additional configuration values within the supplied prefix range. The custom configuration can be supplied via  `spec.kafka.config` with the predefined prefix.
+
+### RemoteLogMetadataManager (RLMM)
+
+For simplicity, the proposal doesn't add customization for RemoteLogMetadataManager. If tiered storage is enabled, the default implementation using `TopicBasedRemoteLogMetadataManager` will be used, and corresponding configuration will be appended. 
+
+Since the default RLMM will create an internal Kafka client to talk to the server, there might be some additional configuration needed, like the SSL related setting. Those config can be put in `.spec.kafka.config` directly using the `rlmm.config.` prefix. Example can be found below.
+
+## Example configuration
+
+The below config define a sample configuration for tiered storage setup, using a `custom` RSM type.
+```
+kafka:
+  tieredStorage:
+    type: custom
+    rsm.className: com.example.kafka.tiered.storage.s3.S3RemoteStorageManager
+    rsm.classPath:  /opt/kafka/plugins/tiered-storage-s3/*
+    rsm.impl.prefix: rsm.config.
+
+  config:
+    ...
+    rlmm.config.remote.log.metadata.common.client.bootstrap.servers: broker-bootstrap.com:9091
+    rlmm.config.remote.log.metadata.common.client.security.protocol: SSL
+```
+
+The configuration above will get translated to the below Kafka broker config:
+```
+remote.log.storage.system.enable: true
+remote.log.storage.manager.class.name: com.example.kafka.tiered.storage.s3.S3RemoteStorageManager
+remote.log.storage.manager.class.path: /opt/kafka/plugins/tiered-storage-s3/*
+remote.log.storage.manager.impl.prefix: rsm.config.
+# The RLMM config will be added by default
+remote.log.metadata.manager.impl.prefix: rlmm.config.
+remote.log.metadata.manager.class.name: org.apache.kafka.server.log.remote.metadata.storage.TopicBasedRemoteLogMetadataManager
+# The RLMM config below are passed down from kafka config spec.
+rlmm.config.remote.log.metadata.common.client.bootstrap.servers: broker-bootstrap.com:9091
+rlmm.config.remote.log.metadata.common.client.security.protocol: SSL
+```
+## Testing strategy
+
+Unit tests will be added to validate all the RLM,RSM, RLMM changes are propagated through.
+Because the RSM and RLMM are typically provided by user and there is no built in support for any third party cloud provider, there won’t be any integration tests covering the cloud provider interaction. It’s the user’s responsibility to ensure the RSM and RLMM works with the intended cloud provider.
+
+For testing purpose, an integration test with LocalTieredStorage implementation can be added to ensure end to end functionality.
+
+## Affected Projects
+
+Strimzi operator will be affected with the change.
```

Comment:
Are we sure that the Topic Operator won't be affected at all?

Reply:
Hi Federico, I am interested to understand more about it. IMO topic Operator is what manages the KafkaTopic resources. In our usage, we didn't touch the KafkaTopic resource. Is there any other areas that could affect topic operator in any way?

Reply:
There are 2 error conditions that are relevant here:

1. TS is enabled on a compacted topic
2. TS is enabled on a topic created with a Kafka version before 2.8.0

I think the Topic Operator should disable these actions and provide warnings (both in KafkaTopic status and Topic Operator logs). We should also investigate if it's possible to detect the 2nd condition.

Reply:
I think these points are worth mentioning in the proposal. That said, I do not think we should handle these in the code and we should leave it to the user's responsibility:
* At least some of them are expected to be removed in the future
* Validating them might be quite complex (for example, IIRC, the compacted topics can be set in the topic itself as well as in the broker defaults)
* Validating this in UTO would capture only a part of the topics and not prevent the users from misconfiguration. So they will still need to be aware of these rules and learn to respect them.

What we probably should mention in the proposal that we should at least one of manually validate the impact on the TO. E.g what happens when tiered storage is disabled and a user tries to create a topic with it? Will it be handled in a reasonable fashion (e.g. if there are errors, they are clear and so on)?

Reply:
Yeah, I suspect that some changes will be required when you start looking into them more closely. I don't think the second point can be verified, so a documentation note should be enough.

Reply:
update the `Validation` section explicitly with the note that at this proposal we won't add complex validation